examples/bbox_annotator.py

```diff
@@ -33,24 +37,34 @@ def create_label_menu(shapes_layer, label_property, labels):
     label_menu = ComboBox(label='text label', choices=labels)
     label_widget = Container(widgets=[label_menu])
 
-    def update_label_menu(event):
+    def update_label_menu():
         """This is a callback function that updates the label menu when
-        the current properties of the Shapes layer change
+        the default features of the Shapes layer change
         """
-        new_label = str(shapes_layer.current_properties[label_property][0])
+        new_label = str(shapes_layer.feature_defaults[label_property][0])
         if new_label != label_menu.value:
             label_menu.value = new_label
 
-    shapes_layer.events.current_properties.connect(update_label_menu)
+    shapes_layer.events.feature_defaults.connect(update_label_menu)
 
-    def label_changed(event):
```

Comment:
I redefined this callback function to not take an `event` argument to avoid the following warning.

```
examples/bbox_annotator.py:69: FutureWarning: 

magicgui 0.4.0 will change the way that callbacks are called.
Instead of a single `Event` instance, with an `event.value` attribute,
callbacks will receive the value(s) directly:

@widget.changed.connect
def my_callback(*args):
    # *args are the value(s) themselves!

To silence this warning you may either provide a callback that has more
or less than 1 parameter.  Or annotate the single parameter as anything
*other* than `Event`, e.g. `def callback(x: int): ...`
For details, see: https://github.com/napari/magicgui/issues/255
  label_menu.changed.connect(label_changed)
```

Instead I just look up the new value as `label_menu.value`. Let me know if I should do something different.

Reply:
I'd say let's have the events carry the information. I prefer callbacks to be self-contained rather than rely on state. So you could silence the warning with just:

```python
def label_changed(value: str):  # yes, annoyingly, pandas doesn't return Categorical, but str, from the series
    ...
```

My *actual* preference would be to use `toolz.curry` to avoid having to make closures using nested function definitions. So:

```python
import toolz as tz

@tz.curry
def set_default_feature(layer, feature_name, value):
    layer.feature_defaults[feature_name] = value
    layer.events.feature_defaults(feature_defaults)
    # optionally we could add feature_names here, containing
    # a list of which feature names have changed defaults

...

label_menu.changed.connect(set_default_feature(shapes_layer, label_property))
# and I would rename `label_property` to `label_feature_name` maybe?
```

What do you think? Currying is definitely advanced, but so is this example so 🤷. Either way I prefer less state even if we can't get to no state.

Reply:
Agreed that using `def label_changed(value: str)` is an improvement because it relies on less captured state in the callback.

I'm less sure about using partials in this example. On the plus side, I'm always a fan of less indentation :) But using them in this example doesn't actually decrease the state we're capturing. Either that state is being captured when defining the nested function or when doing the partial evaluation. If we wanted to reuse those functions (e.g. `set_default_feature`) elsewhere with capturing state, then I see the improvement here. But this example doesn't do that. Maybe you could argue these functions could be copied elsewhere by plugin devs and be more reusable then, but I'm unsure how true that will be in practice.

Anyway, I tried it out and got it working with some wrinkles. Here's [a gist instead](https://gist.github.com/andy-sweet/d63c46359ecfd905877298d921b96ece). As you alluded to, we want to pass the new default values with the `feature_defaults` event. That could just be a reference to `Shapes.feature_defaults`, some subset of its columns, or maybe just a plain `dict`.

As we are adding the `feature_defaults` and `features` events in this PR, now is certainly a good time to think about what values we pass with the event. But I think it's also OK to pass nothing for now and add values later (that won't break people right?).

Reply:
> As we are adding the feature_defaults and features events in this PR, now is certainly a good time to think about what values we pass with the event. But I think it's also OK to pass nothing for now and add values later (that won't break people right?).

Well, it will, if their callback takes no arguments and we start pass arguments, right? If their callback takes `*args` that's true, but I don't see a priori why they would do that.

Reply:
I honestly really like the gist! But I'm wondering why it takes an Event rather than the event value directly? Which I thought is what started this whole discussion? We can bump the magicgui dependency if needed...

Reply:
@tlambert03 do you have thoughts about the two patterns here (nesting vs partials)?

Reply:
> Well, it will, if their callback takes no arguments and we start pass arguments, right?

Yep, I think you're right. I was probably think about [Qt signals and slots](https://doc.qt.io/qt-5/signalsandslots.html), where I think the slot can have fewer arguments than the signal.

Reply:
> But I'm wondering why it takes an Event rather than the event value directly?

Good catch. Here's the error when I try to pass new/changed feature defaults directly rather than the `Event`: https://gist.github.com/andy-sweet/12fc070b25e388232f2ff12a599c31d1

I don't immediately understand the traceback, so maybe you or @tlambert03 can take a look. Note that upgrading the type of the final argument from `Any` to `pd.DataFrame` does not help. My environment has `magicgui==0.3.6`, which is the min required in napari's `setup.cfg`.

Also note that passing the `Event` does not cause the aforementioned warnings. Again, I don't entirely understand why, as the partial does only have one argument, so it seems like the issue I first encountered. Maybe some introspection in the event code is interacting in an interesting way with the partial?

Reply:
> > Well, it will, if their callback takes no arguments and we start pass arguments, right?
> 
> Yep, I think you're right. I was probably thinking about [Qt signals and slots](https://doc.qt.io/qt-5/signalsandslots.html), where I think the slot can have fewer arguments than the signal.

Actually, nope. Our events do seem to behave similarly to Qt, so the callback can have fewer arguments than the event emission. I.e. the following causes `on_features_changed` to be executed without an error.

```python
from napari.utils.events import EmitterGroup, Event

def on_features_changed():
    print('features changed')

group = EmitterGroup(features=Event)
group.features.connect(on_features_changed)
group.features(x=10)
```

Reply:
> I don't immediately understand the traceback, so maybe you or @tlambert03 can take a look.

I pushed the alternative approach in the gist to a branch. See this draft PR on my fork if you want to try to change that to not use `Event` arguments in the callbacks: https://github.com/andy-sweet/napari/pull/4

Reply:
just catching up:
If this is a callback for a magicgui change event (and not an emitter from napari), then, to confirm some of the above: 
1. it's a psygnal based callback (not one from vispy/napari's events.py)
2. like Qt, it _can_ [accept fewer arguments](https://github.com/tlambert03/psygnal#extra-positional-arguments-ignored) than the emitter is providing.  That _is_ intended and well-tested.

... what was the key question about partials here?   the following seems good to me if it meets your needs
```python
@ label_menu.changed.connect
def label_changed(value: str): ...
```

Reply:
sorry, and also, the napari `event.py` also knows whether or not to pass the event object to the callback too (see the `_callback_pass_event` store in event.py)

Reply:
> sorry, and also, the napari event.py also knows whether or not to pass the event object to the callback too (see the _callback_pass_event store in event.py)

Fascinating! Has that always been the case? I remember early on futzing around with lambdas to make sure that the event emission and the callback signatures would match, with `lambda ev: func()`

> ... what was the key question about partials here?

```python
@label_menu.changed.connect
def label_changed(value: str): ...
```

This definition is nested within `label_menu` so that it has access to the viewer and layers, which are then used in a closure. My proposal is to un-nest it and pass those as arguments, then use partials (either with toolz.curry or functools.partial) to connect.

```python
def label_changed(viewer, layer, value: str):
    ...

...

label_menu.changed.connect(partial(label_changed, viewer, layer))
```

@andy-sweet 

> My environment has magicgui==0.3.6, which is the min required in napari's setup.cfg

We can potentially bump that dependency. I think this might need a newer magicgui.

Reply:
> If this is a callback for a magicgui change event (and not an emitter from napari),

I hate extending this thread, but as in my example above (and in this PR), we also get the described behavior for napari events/emitters. Is that intended?

Reply:
yes, see https://github.com/napari/napari/pull/4218#discussion_r830659211

Reply:
> This definition is nested within label_menu so that it has access to the viewer and layers, which are then used in a closure. My proposal is to un-nest it and pass those as arguments, then use partials (either with toolz.curry or functools.partial) to connect.

I don't have a strong preference here either.  I think the main concern here is holding a strong reference, but in this case, I suspect _both_ the closure and the partial would hold a strong reference.  It's just an example though, so let's not overthink it :joy: they both work.